services/agents/channel/channel.go

```diff
@@ -272,31 +289,50 @@ func (c *Channel) subscribe(id uint32) chan agentpb.AgentResponsePayload {
 	_, ok := c.responses[id]
 	if ok {
 		// it is possible only on lastSentRequestID wrap around, and we can't recover from that
-		logger.Get(c.s.Context()).Panicf("Already have subscriber for ID %d.", id)
+		c.l.Panicf("Already have subscriber for ID %d.", id)
 	}
 
 	c.responses[id] = ch
 	c.rw.Unlock()
 	return ch
 }
 
-func (c *Channel) publish(id uint32, resp agentpb.AgentResponsePayload) {
+func (c *Channel) removeResponseChannel(id uint32) chan Response {
 	c.rw.Lock()
 	if c.responses == nil { // Channel is closed, no more publishing
 		c.rw.Unlock()
```

Comment:
looks like we can just defer unlock since we don't close it here anymore

Reply:
done